tensorly/decomposition/candecomp_parafac.py

```diff
@@ -247,21 +315,38 @@ def parafac(tensor, rank, n_iter_max=100, init='svd', svd='numpy_svd',\
 
             factors[mode] = factor
 
-        if tol:
-            if sparsity:
-                low_rank_component = kruskal_to_tensor((weights, factors))
-                sparse_component = sparsify_tensor(tensor - low_rank_component, sparsity)
-                
-                unnorml_rec_error = tl.norm(tensor - low_rank_component - sparse_component, 2)
-            else:
-                # ||tensor - rec||^2 = ||tensor||^2 + ||rec||^2 - 2*<tensor, rec>
-                factors_norm = kruskal_norm((weights, factors))
+        # Calculate the current unnormalized error if we need it
+        if tol or (linesearch and iteration % 2 == 0):
+            unnorml_rec_error, tensor, norm_tensor = error_calc(tensor, norm_tensor, weights, factors, sparsity, mask, mttkrp)
+        else:
+            if mask is not None:
+                tensor = tensor*mask + tl.kruskal_to_tensor((weights, factors), mask=1-mask)
 
-                # mttkrp and factor for the last mode. This is equivalent to the
-                # inner product <tensor, factorization>
-                iprod = tl.sum(tl.sum(mttkrp*factor, axis=0)*weights)
-                unnorml_rec_error = tl.sqrt(tl.abs(norm_tensor**2 + factors_norm**2 - 2*iprod))
-                
+        # Start line search if requested. Doesn't yet work for fixed modes.
+        if linesearch and iteration % 2 == 0:
+            jump = iteration ** (1.0/3.0)
+
+            weights_diff = weights - weights_last
+            factors_diff = [factors[ii] - factors_last[ii] for ii in range(tl.ndim(tensor))]
+
+            # Keep stepping while we're successful.
+            new_weights = weights + jump * weights_diff
```

Comment:
I think Bro jumps from `weights_last` instead of `weights`, because the jump here is essentially (1 + jump) per iteration. However, the error calculation for a jump of 1.0 is much cheaper since we already have `mttkrp`. I'm interested in what you think.